src/hypertable_with_clause.c

```diff
@@ -0,0 +1,284 @@
+	chunk_target_size       TEXT = NULL,
+	chunk_sizing_func       REGPROC = '_timescaledb_internal.calculate_chunk_interval'::regproc,
+	time_partitioning_func  REGPROC = NULL,
+	replication_factor      INTEGER = NULL,
+	data_nodes              NAME[] = NULL,
+	distributed             BOOLEAN = NULL
+*/
+
+static const WithClauseDefinition hypertable_with_clause_def[] = {
+		[HypertableEnabled] = {
+			.arg_name = "hypertable",
+			.type_id = BOOLOID,
+			.default_val = BoolGetDatum(true),
+		},
+		[HypertableTimeColumn] = {
+			.arg_name = "time_column",
+			.type_id = NAMEOID,
+		},
+		[HypertablePartitioningColumn] = {
+			.arg_name = "partitioning_column",
+			.type_id = NAMEOID,
+		},
+		[HypertableNumberPartitions] = {
+			.arg_name = "number_partitions",
+			.type_id = INT4OID,
+		},
+		[HypertableAssociatedSchemaName] = {
+			.arg_name = "associated_schema_name",
+			.type_id = NAMEOID,
+		},
+		[HypertableAssociatedTablePrefix] = {
+			.arg_name = "associated_table_prefix",
+			.type_id = NAMEOID,
+		},
+		[HypertableChunkTimeInterval] = {
+			.arg_name = "chunk_time_interval",
+			.type_id = TEXTOID,
+		},
+		[HypertableCreateDefaultIndexes] = {
+			.arg_name = "create_default_indexes",
+			.type_id = BOOLOID,
+		},
+		[HypertableIfNotExists] = {
+			.arg_name = "if_not_exists",
+			.type_id = BOOLOID,
+			.default_val = BoolGetDatum(true),
+		},
+		[HypertablePartitioningFunc] = {
+			.arg_name = "partitioning_func",
+			.type_id = REGPROCOID,
+		},
+		[HypertableMigrateData] = {
+			.arg_name = "migrate_data",
+			.type_id = BOOLOID,
+		},
+		[HypertableChunkTargetSize] = {
+			.arg_name = "chunk_target_size",
+			.type_id = TEXTOID,
+		},
+		[HypertableChunkSizingFunc] = {
+			.arg_name = "chunk_sizing_func",
+			.type_id = REGPROCOID,
+		},
+		[HypertableTimePartitioningFunc] = {
+			.arg_name = "time_partitioning_func",
+			.type_id = REGPROCOID,
+		},
+		[HypertableReplicationFactor] = {
+			.arg_name = "replication_factor",
+			.type_id = INT4OID,
+		},
+		[HypertableDataNodes] = {
+			.arg_name = "data_nodes",
+			.type_id = NAMEARRAYOID,
+		},
+		[HypertableDistributed] = {
+			.arg_name = "distributed",
+			.type_id = BOOLOID,
+		},
+		[HypertableCompressEnabled] = {
+			.arg_name = "compress",
+			.type_id = BOOLOID,
+			.default_val = BoolGetDatum(false),
+		},
+		[HypertableCompressSegmentBy] = {
+			 .arg_name = "compress_segmentby",
+			 .type_id = TEXTOID,
+		},
+		[HypertableCompressOrderBy] = {
+			 .arg_name = "compress_orderby",
+			 .type_id = TEXTOID,
+		},
+		[HypertableCompressChunkTimeInterval] = {
+			 .arg_name = "compress_chunk_time_interval",
+			 .type_id = INTERVALOID,
+		}
+};
+
+static Oid
+get_chunk_sizing_func_oid(void)
+{
+	Oid argtype[] = { INT4OID, INT8OID, INT8OID };
+	return ts_get_function_oid("calculate_chunk_interval", "_timescaledb_internal", 3, argtype);
+}
+
+static Oid
+get_create_hypertable_func_oid(void)
+{
+	Oid argtype[] = { REGCLASSOID,	 NAMEOID,	 NAMEOID, INT4OID,		NAMEOID, NAMEOID,
```

Comment:
the current approach seem to build on the concept of having an enum stuff with the same oid; 
could you assert / make sure the type ids are aligned here and in the enum somehow?

...or collect the type ids at this method's callsite and do the get based on that (no need for this method) - so that you don't need to list these here...and if there is an issue there will be just no function to be called...

Reply:
Yeah, I understand your point. Since PostgreSQL supports polymorphic functions it is using argument and name to find exact match. Essentially it is possible to have several functions with the same name but with different arguments. It will produce an error if there are no function found during out tests, which will get the idea that we forgot to update the function. 

The problem here is that we also need those types for storage options convertion and we can't completely automate it, since there are some exceptions, like `chunk_time_interval`, etc